libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/AppMetadata.kt

```diff
@@ -0,0 +1,77 @@
+package com.mapbox.navigation.core.telemetry.events
+
+/**
+ * Custom metadata that can be used to associate app state with feedback events in the telemetry pipeline.
+ */
+data class AppMetadata constructor(
+    /**
+     * Name of the application. Value should be non-empty.
+     */
+    val name: String, // min length 1
+
+    /**
+     * Version of the application. Value should be non-empty.
+     */
+    val version: String, // min length 1
+
+    /**
+     * User ID relevant for the application context.
+     */
+    val userId: String?,
+
+    /**
+     * Session ID relevant for the application context.
+     */
+    val sessionId: String?
+) {
+
+    /**
+     * Get a builder to customize a subset of current options.
+     */
+    fun toBuilder() = Builder(name, version).apply {
+        userId(userId)
+        sessionId(sessionId)
+    }
+
+    /**
+     * [AppMetadata] builder.
+     */
+    class Builder(
+        /**
+         * Name of the application. Value should be non-empty.
+         */
+        private val name: String, // min length 1
+
+        /**
+         * Version of the application. Value should be non-empty.
+         */
+        private val version: String // min length 1
```

Comment:
give them default values outside of constructor. you can run validation in the `build()` function

``` kotlin
        /**
         * Builds new [AppMetadata] instance.
         */
        fun build() = {
              check(name.isNotEmpty) { "non empty name is required" }
              return AppMetadata(...)
```

Reply:
Why not require the argument in the constructor? Are there any additional benefits? This also refs the concerns raised by @cafesilencio in https://github.com/mapbox/mapbox-navigation-android/issues/2709#issuecomment-658848477.

Reply:
yeah not disagreeing with the comment. @cafesilencio is saying the builder should not fail at runtime because you didn't supply it. my comment is assuming that `name` was set initially with a default value. and the developer set name to empty.

or the developer is using a value source that breaks the assumptions we'll have downstream. the "min length 1" comment does not actually prevent someone from building the object with an empty string.